data_model/src/events/time.rs

```diff
@@ -0,0 +1,110 @@
+//! Time event and filter
+
+use core::{ops::Range, time::Duration};
+
+use super::*;
+
+/// Special event that is emitted when `WSV` is ready for handling time-triggers
+///
+/// Contains time interval which is used to identify time-triggers to be executed
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Decode, Encode, IntoSchema)]
+pub struct Event(pub Interval);
+
+/// Filters time-events and allows only ones which time interval contains
+#[derive(
+    Debug,
+    PartialOrd,
+    Ord,
+    PartialEq,
+    Eq,
+    Clone,
+    Copy,
+    Decode,
+    Encode,
+    IntoSchema,
+    Hash,
+    Serialize,
+    Deserialize,
+)]
+pub struct EventFilter(pub Occurrence);
+
+impl EventFilter {
+    /// Check if `event` matches filter
+    ///
+    /// Returns `true`, if `event` matches filter and `false` if not
+    pub fn matches(&self, event: &Event) -> bool {
+        match &self.0 {
+            Occurrence::Every(interval) => {
+                let time = interval.since + interval.length;
+                Range::from(event.0).contains(&time)
+            }
```

Comment:
Indeed that's confusing and there was a bug I forgot about. I've fixed it.

`interval.since` should be updated every time time-trigger is executed. It should be set to the event time (block timestamp). So that calculations here will be alright.

This thing is needed to prevent double execution if `wsv.apply()` will happen to close

Reply:
How do we get the first time the trigger was run? We don't want them to drift.

Reply:
Closing this, cause it's not releavant for the new realization